site/en/docs/extensions/mv3/automated-testing/index.md

```diff
@@ -0,0 +1,149 @@
+
+| Library                | Guidance                                                                                                                              |
+|------------------------|:-------------------------------------------------------------------------------------------------------------------------------------:|
+| Puppeteer / Playwright | See Chrome Extensions ([Puppeteer][puppeteer-testing] / [Playwright][playwright-testing]).                                            |
+| Selenium               | Use the [ChromeOptions][selenium-chromeoptions] object to load extensions. More information is available [here][selenium-extensions]. |
+| WebDriverIO            | See [Web Extension Testing][webdriverio-testing].                                                                                     |
+
+### Headless Chrome
+
+When running tests as part of an automated workflow, it is often necessary to load your extension on a machine that does not have a screen. Chrome's [new headless][new-headless] mode allows Chrome to be run in an unattended environment like this. Simply start Chrome with the `--headless=new` flag (headless currently defaults to "old", which does not support loading extensions). Depending on the automation tool you choose there may be a setting which adds the flag for you automatically.
+
+### Setting an extension ID
+
+It is often desirable to have a fixed extension ID in tests. This makes a number of common tasks easier such as allow-listing the extension's origin on a server it needs to communicate with, or opening extension pages within tests. To do this, follow the steps under [Keeping a consistent extension ID][consistent-id].
+
+### Testing extension pages
+
+Extension pages can be accessed using their corresponding URL, e.g `chrome-extension://<id>/index.html`. Use the normal navigation methods available in your automation tool of choice to navigate to these URLs.
+
+### Testing an extension popup
+
+Opening an extension popup in the context of another page is not currently possible. Instead, open the popup's URL in a new tab. If your popup uses the active tab, consider implementing an override where a tab ID can be passed explicitly to your popup. For example:
+
+```js
+const URL_PARAMS  = new URLSearchParams(window.location.search);
+
+async function getActiveTab() {
+  // Open popup.html?tab=5 to use tab ID 5, etc.
+  if (URL_PARAMS.has("tab")) {
+    return parseInt(URL_PARAMS.get("tab"));
+  }
+
+  const tabs = await chrome.tabs.query({
+    active: true,
+    currentWindow: true
+  });
+
+  return tabs[0];
+}
+```
+
+### Inspecting extension state
+
+It is generally best practice to avoid accessing internal state in an integration test, and instead you should base your tests on what is visible to the user. However, it can sometimes be desirable to directly access data from the extension. In these cases, consider executing code in the context of an extension page.
```

Comment:
```suggestion
To avoid that test fail, when you change the internal behavior of your extensions, it is generally best practice to avoid accessing internal state in an integration test. Instead you should base your tests on what is visible to the user. However, it can sometimes be desirable to directly access data from the extension. In these cases, consider executing code in the context of an extension page.
```